src/shim.c

```
@@ -703,6 +704,9 @@ handle_proxy_message(struct cc_shim *shim)
 
 	fr = read_frame(shim);
 	if ( !fr) {
+		if (! reconnect_to_proxy(shim)) {
```

You're making lots of assumption by calling this after you got a `fr == NULL`. Indeed, looking at `read_frame()` function, there are several reasons why we could get an error, and they are not all related to the proxy disconnection.

---

@sboeuf, thanks for the review.

Okay. Re-connection when header length or header version is wrong doesn't make much sense. If I change for these cases instead of returning false to abort()? Would it make sense for you?

---

If I understand you well, you want to change the `return false` with `abort()` in case you get header length or header version which is wrong ? Because you consider those cases like real issues and we should just terminate the shim, right ?

---

Sorry for the delayed response. 
So read_frame() can return NULL in case when read_wire_data() fails --> we re-connect.
Also read_frame() return NULL in cases:
 * shim == NULL or shim->proxy_sock_fd < 0
 * fr->header.header_len < MIN_HEADER_WORD_SIZE
 * fr->header.version != PROXY_VERSION

My proposition is to handle these cases the same as failures with allocation --> calling abort(). Doing this way will make read_frame() return NULL only in cases when read_wire_date() failed.